### PR TITLE
Android P FileStore flush fix

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Some unhandled error reports aren't being sent immediately on Android P and in situations with no connectivity.

Internally `FileStore` retains a `Collection` of files which have previously been found for a request, to prevent duplicate reports being sent. This puts the onus on the calling code to cancel or delete the files once the request has been completed. 

This changeset ensures that these files are only requested once when calling `flushOnLaunch`, then cancels all files after delivery has taken place (successfully or not), which resets the `FileStore` state.

After this, `flushAsync` is then called, which will flush any remaining reports that have not yet been delivered.

### Changed

`FileStore` (see above description)

## Tests

Manual test on Android P simulator of unhandled crashes. Mazerunner scenario that fails prior to the applied fix.

### Alternative Approaches

The delivery/flushing code is fairly complex, it may be worth rethinking our implementation at some future point.

### Linked issues

#318 

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [x] Consistency across platforms for structures or concepts added or modified
- [x] Consistency between the changeset and the goal stated above
- [x] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [x] Usage friction - is the proposed change in usage cumbersome or complicated?
- [x] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [x] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [x] Thoroughness of added tests and any missing edge cases
- [x] Idiomatic use of the language

### CODE REVIEW

**Comment by fractalwrench** [comment]
I've updated the mazerunner scenario to throw exceptions rather than writing errors to the store.

I've also made a slight adjustment to the `ErrorStore` logic so that if `config.launchCrashThresholdMs == 0`, all reports will be flushed async, as launch crash delivery behaviour is disabled. This addresses an internal ticket.

File: mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/StartupCrashFlushScenario.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+internal class StartupCrashFlushScenario(config: Configuration,
+                                         context: Context) : Scenario(config, context) {
+
+    override fun run() {
+        super.run()
+        disableAllDelivery()
+
+        if ("StartupCrash" == eventMetaData) {
+            config.launchCrashThresholdMs = Long.MAX_VALUE
```
Comment (snmaynard) [inline mazerunner/src/main/java/com/bugsnag/android/mazerunner/scenarios/StartupCrashFlushScenario.kt]: I think this needs some rethinking. I think you dont want to set `config.launchCrashThresholdMs` in here. Also both of these crashes will count as launch crashes on subsequent boots.

**Comment by fractalwrench** [comment]
Updated the docs to make it a bit clearer what this scenario is doing, and also updated mazerunner to disable delivery before Bugsnag is initialised.

**Comment by kattrali** [comment]
Reworked the test to better illustrate the user scenario here, and introduced a delay to separate the launch crash from a regular one.
